File: invariant/invariant-gates/src/receipt.rs
// invariant-gates/src/receipt.rs
// ─────────────────────────────────────────────────────────────────
// INVARIANT receipt data structures.
// These are the canonical types.  Python code converts to/from JSON.
// The 136-byte serialised form is what travels over Axon/Dendrite.
// ─────────────────────────────────────────────────────────────────

use serde::{Deserialize, Serialize};

/// Fixed-size receipt — exactly 136 bytes on the wire.
///
/// Field layout (big-endian):
///   agent_id        32 bytes   SHA-256(hotkey || model_hash || reg_block)
///   model_hash      32 bytes   SHA-256(model identifier / weights)
///   execution_hash  32 bytes   SHA-256(task_input || output || tempo_id || ts)
///   counter          8 bytes   monotonic uint64
///   digest          32 bytes   SHA-256(agent_id || model_hash || execution_hash || counter)
///   ─────────────────────────
///   TOTAL          136 bytes
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Receipt {
    pub agent_id: [u8; 32],
    pub model_hash: [u8; 32],
    pub execution_hash: [u8; 32],
    pub counter: u64,
    pub digest: [u8; 32],

    // ── Metadata (not included in digest) ───────────────────────
    #[serde(default)]
    pub version: u8,
    #[serde(default)]
    pub timestamp: f64,
    #[serde(default)]
    pub tempo_id: u64,
}

impl Receipt {
    /// Serialise to exactly 136 bytes (digest fields only).
    pub fn to_bytes(&self) -> [u8; 136] {
        let mut buf = [0u8; 136];
        buf[0..32].copy_from_slice(&self.agent_id);
        buf[32..64].copy_from_slice(&self.model_hash);
        buf[64..96].copy_from_slice(&self.execution_hash);
        buf[96..104].copy_from_slice(&self.counter.to_be_bytes());
        buf[104..136].copy_from_slice(&self.digest);
        buf
    }

    /// Deserialise from 136 bytes.
    pub fn from_bytes(b: &[u8; 136]) -> Self {
        let mut agent_id = [0u8; 32];
        let mut model_hash = [0u8; 32];
        let mut execution_hash = [0u8; 32];
        let mut digest = [0u8; 32];

        agent_id.copy_from_slice(&b[0..32]);
        model_hash.copy_from_slice(&b[32..64]);
        execution_hash.copy_from_slice(&b[64..96]);
        let counter = u64::from_be_bytes(b[96..104].try_into().unwrap());
        digest.copy_from_slice(&b[104..136]);

        Receipt {
            agent_id,
            model_hash,
            execution_hash,
            counter,
            digest,
            version: 1,
            timestamp: 0.0,
            tempo_id: 0,
        }
    }
}

/// JSON-serialisable form used for transport over Axon (hex strings).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReceiptJson {
    pub version: u8,
    pub agent_id: String,       // hex
    pub model_hash: String,     // hex
    pub execution_hash: String, // hex
    pub counter: u64,
    pub digest: String, // hex
    pub timestamp: f64,
    pub tempo_id: u64,
}

impl ReceiptJson {
    pub fn from_receipt(r: &Receipt) -> Self {
        Self {
            version: r.version,
            agent_id: hex::encode(r.agent_id),
            model_hash: hex::encode(r.model_hash),
            execution_hash: hex::encode(r.execution_hash),
            counter: r.counter,
            digest: hex::encode(r.digest),
            timestamp: r.timestamp,
            tempo_id: r.tempo_id,
        }
    }

    pub fn to_receipt(&self) -> Result<Receipt, hex::FromHexError> {
        let decode32 = |s: &str| -> Result<[u8; 32], hex::FromHexError> {
            let v = hex::decode(s)?;
            let mut arr = [0u8; 32];
            arr.copy_from_slice(&v);
            Ok(arr)
        };
        Ok(Receipt {
            agent_id: decode32(&self.agent_id)?,
            model_hash: decode32(&self.model_hash)?,
            execution_hash: decode32(&self.execution_hash)?,
            counter: self.counter,
            digest: decode32(&self.digest)?,
            version: self.version,
            timestamp: self.timestamp,
            tempo_id: self.tempo_id,
        })
    }
}

/// Outcome of the four-gate verification pass.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum GateResult {
    Pass,
    Gate1AgentNotAuthorized,
    Gate2ModelNotApproved,
    Gate3ReplayDetected,
    Gate4DigestMismatch,
    ParseError(String),
}

impl GateResult {
    pub fn is_pass(&self) -> bool {
        matches!(self, GateResult::Pass)
    }

    /// Human-readable code (used by Python side for OAP violation mapping).
    pub fn code(&self) -> &'static str {
        match self {
            GateResult::Pass => "PASS",
            GateResult::Gate1AgentNotAuthorized => "GATE1_AGENT_NOT_AUTHORIZED",
            GateResult::Gate2ModelNotApproved => "GATE2_MODEL_NOT_APPROVED",
            GateResult::Gate3ReplayDetected => "GATE3_REPLAY_DETECTED",
            GateResult::Gate4DigestMismatch => "GATE4_DIGEST_MISMATCH",
            GateResult::ParseError(_) => "PARSE_ERROR",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_bytes() {
        let r = Receipt {
            agent_id: [1u8; 32],
            model_hash: [2u8; 32],
            execution_hash: [3u8; 32],
            counter: 42,
            digest: [4u8; 32],
            version: 1,
            timestamp: 0.0,
            tempo_id: 100,
        };
        let bytes = r.to_bytes();
        assert_eq!(bytes.len(), 136);
        let r2 = Receipt::from_bytes(&bytes);
        assert_eq!(r.agent_id, r2.agent_id);
        assert_eq!(r.counter, r2.counter);
        assert_eq!(r.digest, r2.digest);
    }

    #[test]
    fn round_trip_json() {
        let r = Receipt {
            agent_id: [0xABu8; 32],
            model_hash: [0xCDu8; 32],
            execution_hash: [0xEFu8; 32],
            counter: 999,
            digest: [0x12u8; 32],
            version: 1,
            timestamp: 1_700_000_000.0,
            tempo_id: 50,
        };
        let j = ReceiptJson::from_receipt(&r);
        let r2 = j.to_receipt().unwrap();
        assert_eq!(r.agent_id, r2.agent_id);
        assert_eq!(r.counter, r2.counter);
    }
}

File: sdk/invariant-miner/invariant_miner/receipt.py
"""
invariant_miner/receipt.py
=========================
Receipt dataclass — the 136-byte execution proof produced by every
INVARIANT miner after task execution.

The receipt is submitted alongside the task output to the validator.
The validator runs all four gates against it before scoring output quality.

Wire format (136 bytes, big-endian):
    agent_id        [0:32]   SHA-256(hotkey || model_hash || reg_block)
    model_hash      [32:64]  SHA-256(model_identifier_string)
    execution_hash  [64:96]  SHA-256(task_input || output || tempo_id || timestamp)
    counter         [96:104] uint64, strictly monotonic
    digest          [104:136] SHA-256(agent_id || model_hash || exec_hash || counter)

The JSON form (for Axon/Dendrite transport) uses hex strings for byte fields.

Usage
-----
    from invariant_miner import build_receipt

    receipt = build_receipt(
        agent_id="<64-char hex>",
        model_identifier="my-model-v1",
        task_input="What is 2+2?",
        output="4",
        counter=1,
        tempo_id=100,
        timestamp=time.time(),
    )

    # Send over the wire
    json_str = receipt.to_json()

    # Reconstruct on the validator side
    receipt2 = Receipt.from_json(json_str)
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass, field
from typing import Any, Dict

_WIRE_SIZE = 136  # bytes


@dataclass(frozen=True)
class Receipt:
    """An INVARIANT execution receipt.

    All byte fields are stored as lowercase hex strings for easy
    JSON serialisation and human inspection.

    Attributes
    ----------
    agent_id : str
        64-char hex.  SHA-256(hotkey || model_hash || registration_block).
        Identifies the miner uniquely; hotkey-bound so it cannot be forged.
    model_hash : str
        64-char hex.  SHA-256(model_identifier_string).
        Must appear in the validator's approved model registry (Gate 2).
    execution_hash : str
        64-char hex.  SHA-256(task_input || output || tempo_id || timestamp).
        Cryptographically binds the specific input/output pair and tempo.
        Cannot be produced without having computed the output.
    counter : int
        Monotonic uint64.  Must be strictly greater than the last confirmed
        counter for this agent (Gate 3 — replay protection).
        Increment by at least 1 each tempo.
    digest : str
        64-char hex.  SHA-256(agent_id || model_hash || execution_hash || counter).
        Integrity seal over all four core fields (Gate 4).
    version : int
        Receipt format version.  Current: 1.
    timestamp : float
        Unix timestamp (seconds) at receipt creation.  Informational only —
        not included in the digest.
    tempo_id : int
        Bittensor tempo identifier.  Included in execution_hash so receipts
        cannot be reused across tempos (cross-tempo caching attack).
    """

    agent_id: str  # 64-char hex
    model_hash: str  # 64-char hex
    execution_hash: str  # 64-char hex
    counter: int  # uint64, monotonic
    digest: str  # 64-char hex
    version: int = 1
    timestamp: float = 0.0
    tempo_id: int = 0

    # ------------------------------------------------------------------
    # Validation
    # ------------------------------------------------------------------

    def __post_init__(self) -> None:
        for name in ("agent_id", "model_hash", "execution_hash", "digest"):
            val = getattr(self, name)
            if len(val) != 64:
                raise ValueError(
                    f"Receipt.{name} must be 64 hex characters (32 bytes), "
                    f"got {len(val)}"
                )
            try:
                bytes.fromhex(val)
            except ValueError as e:
                raise ValueError(f"Receipt.{name} is not valid hex: {e}") from e
        if self.counter < 0:
            raise ValueError(f"Receipt.counter must be >= 0, got {self.counter}")
        if self.tempo_id < 0:
            raise ValueError(f"Receipt.tempo_id must be >= 0, got {self.tempo_id}")

    # ------------------------------------------------------------------
    # Serialisation — JSON (transport format)
    # ------------------------------------------------------------------

    def to_dict(self) -> Dict[str, Any]:
        """Return the receipt as a plain dict (JSON-serialisable).

        This is the form transmitted over Axon/Dendrite.

        Returns:
            Dict with string hex fields for all byte arrays and
            int/float for scalar fields.
        """
        return {
            "version": self.version,
            "agent_id": self.agent_id,
            "model_hash": self.model_hash,
            "execution_hash": self.execution_hash,
            "counter": self.counter,
            "digest": self.digest,
            "timestamp": self.timestamp,
            "tempo_id": self.tempo_id,
        }

    def to_json(self) -> str:
        """Serialise the receipt to a compact JSON string.

        Returns:
            Compact JSON string — safe to embed in a Bittensor synapse field.
        """
        return json.dumps(self.to_dict(), separators=(",", ":"))

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Receipt":
        """Reconstruct a Receipt from a plain dict.

        Args:
            d: A dict previously produced by :meth:`to_dict` or decoded
               from the JSON transport form.

        Returns:
            A validated Receipt instance.

        Raises:
            KeyError:   If a required field is missing.
            ValueError: If any field fails validation.
        """
        return cls(
            agent_id=d["agent_id"],
            model_hash=d["model_hash"],
            execution_hash=d["execution_hash"],
            counter=int(d["counter"]),
            digest=d["digest"],
            version=int(d.get("version", 1)),
            timestamp=float(d.get("timestamp", 0.0)),
            tempo_id=int(d.get("tempo_id", 0)),
        )

    @classmethod
    def from_json(cls, json_str: str) -> "Receipt":
        """Deserialise a Receipt from a JSON string.

        Args:
            json_str: A JSON string previously produced by :meth:`to_json`.

        Returns:
            A validated Receipt instance.

        Raises:
            json.JSONDecodeError: If the string is not valid JSON.
            KeyError:   If a required field is missing.
            ValueError: If any field fails validation.
        """
        return cls.from_dict(json.loads(json_str))

    # ------------------------------------------------------------------
    # Serialisation — binary wire format (136 bytes)
    # ------------------------------------------------------------------

    def to_bytes(self) -> bytes:
        """Serialise to the canonical 136-byte binary wire format.

        Layout (big-endian):
            [0:32]   agent_id
            [32:64]  model_hash
            [64:96]  execution_hash
            [96:104] counter (uint64 BE)
            [104:136] digest

        Returns:
            Exactly 136 bytes.
        """
        buf = bytearray(_WIRE_SIZE)
        buf[0:32] = bytes.fromhex(self.agent_id)
        buf[32:64] = bytes.fromhex(self.model_hash)
        buf[64:96] = bytes.fromhex(self.execution_hash)
        buf[96:104] = struct.pack(">Q", self.counter)
        buf[104:136] = bytes.fromhex(self.digest)
        return bytes(buf)

    @classmethod
    def from_bytes(cls, data: bytes | bytearray) -> "Receipt":
        """Deserialise from the canonical 136-byte binary wire format.

        Args:
            data: Exactly 136 bytes in the canonical layout.

        Returns:
            A validated Receipt instance (version=1, timestamp=0.0, tempo_id=0).

        Raises:
            ValueError: If data is not exactly 136 bytes.
        """
        if len(data) != _WIRE_SIZE:
            raise ValueError(f"Expected exactly {_WIRE_SIZE} bytes, got {len(data)}")
        return cls(
            agent_id=data[0:32].hex(),
            model_hash=data[32:64].hex(),
            execution_hash=data[64:96].hex(),
            counter=struct.unpack(">Q", data[96:104])[0],
            digest=data[104:136].hex(),
            version=1,
            timestamp=0.0,
            tempo_id=0,
        )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def short_id(self) -> str:
        """Return the first 12 characters of agent_id for log messages."""
        return self.agent_id[:12]

    def short_digest(self) -> str:
        """Return the first 12 characters of digest for log messages."""
        return self.digest[:12]

    def __repr__(self) -> str:
        return (
            f"Receipt(agent={self.short_id()}…, "
            f"counter={self.counter}, "
            f"tempo={self.tempo_id}, "
            f"digest={self.short_digest()}…)"
        )
